File: engine/argot/scoring/scorers/fall_through_guards.py
"""Fall-through-guard count primitive.

Scalar: count of ``if_statement`` nodes that appear (by source position)
before the first ``return_statement`` inside each ``function_definition`` body.
Per-cluster baseline is ``(mean, std)`` of the per-file average guard count
across cluster files.  Score-time contribution is a tail-z penalty clipped at
``cluster_bonus_clip``.

"Before return" definition:
  Within a ``function_definition`` subtree, count all ``if_statement`` nodes
  whose start_byte is strictly less than the start_byte of the earliest
  ``return_statement`` in that same subtree.  If the function contains no
  ``return_statement``, guard count is 0 (no guards before a return that
  does not exist).

Ramp curve:
  contribution = min(cluster_bonus_clip, max(0.0, abs(tail_z) - 1.0))
  Two-sided: catches average guard counts anomalously HIGH or LOW versus
  cluster baseline.  Ramps linearly from 0 at |z|=1σ up to cluster_bonus_clip
  at |z| = 1 + cluster_bonus_clip σ (=6σ for the default clip of 5.0).
"""

from __future__ import annotations

import statistics
from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from pathlib import Path

from tree_sitter import Node

from argot.scoring.filters.typicality import _PY_PARSER, _TS_PARSER
from argot.scoring.scorers.shape_primitive import Language

# ---------------------------------------------------------------------------
# Baseline payload
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class _GuardCountBaseline:
    mean: float
    std: float


# ---------------------------------------------------------------------------
# AST helpers
# ---------------------------------------------------------------------------

# tree-sitter node kind for function definitions (both grammars).
# Python: function_definition; TypeScript: function_declaration.
_PY_FUNC_TYPE = "function_definition"
_TS_FUNC_TYPE = "function_declaration"

_IF_TYPE = "if_statement"
_RETURN_TYPE = "return_statement"


def _walk(root: Node) -> Iterator[Node]:
    """Pre-order DFS over all nodes in the subtree rooted at *root*."""
    stack: list[Node] = [root]
    while stack:
        node = stack.pop()
        yield node
        stack.extend(reversed(node.children))


def _guards_before_return(func_node: Node) -> int:
    """Count ``if_statement`` nodes before the first ``return_statement`` in *func_node*.

    Returns 0 when the function contains no ``return_statement`` (no guards
    can precede a return that does not exist).
    """
    first_return_byte: int | None = None
    if_bytes: list[int] = []

    for node in _walk(func_node):
        if node.type == _RETURN_TYPE:
            if first_return_byte is None or node.start_byte < first_return_byte:
                first_return_byte = node.start_byte
        elif node.type == _IF_TYPE:
            if_bytes.append(node.start_byte)

    if first_return_byte is None:
        return 0
    return sum(1 for b in if_bytes if b < first_return_byte)


def _file_avg_guards(source: str, language: Language) -> float | None:
    """Return mean guard-count per function for *source*, or None if no functions."""
    parser = _PY_PARSER if language == "python" else _TS_PARSER
    func_type = _PY_FUNC_TYPE if language == "python" else _TS_FUNC_TYPE
    try:
        tree = parser.parse(source.encode("utf-8"))
    except Exception:  # noqa: BLE001
        return None

    guard_counts: list[int] = []
    for node in _walk(tree.root_node):
        if node.type == func_type:
            guard_counts.append(_guards_before_return(node))

    del tree
    if not guard_counts:
        return None
    return statistics.mean(guard_counts)


# ---------------------------------------------------------------------------
# Primitive
# ---------------------------------------------------------------------------

_MIN_VALID_FILES = 3


@dataclass
class FallThroughGuards:
    """Fall-through-guard count primitive.

    Implements ``ShapePrimitive[_GuardCountBaseline]``.
    """

    name: str = "fall_through_guards"
    min_cluster_size: int = 10
    cluster_bonus_clip: float = 5.0

    # ------------------------------------------------------------------
    # Fit
    # ------------------------------------------------------------------

    def fit_cluster_baseline(
        self,
        cluster_files: Iterable[tuple[Path, str]],
        language: Language,
    ) -> _GuardCountBaseline | None:
        """Compute mean ± std of per-file average guard count across cluster files.

        Returns ``None`` if fewer than ``_MIN_VALID_FILES`` files have at least
        one ``function_definition`` (insufficient sample).
        """
        avgs: list[float] = []
        for _path, source in cluster_files:
            avg = _file_avg_guards(source, language)
            if avg is not None:
                avgs.append(avg)
        if len(avgs) < _MIN_VALID_FILES:
            return None
        mean = statistics.mean(avgs)
        std = statistics.pstdev(avgs) if len(avgs) > 1 else 0.0
        return _GuardCountBaseline(mean=mean, std=std)

    # ------------------------------------------------------------------
    # Score
    # ------------------------------------------------------------------

    def score(
        self,
        hunk_content: str,
        *,
        baseline: _GuardCountBaseline | None,
        cluster_size: int,
    ) -> float:
        """Return tail-z contribution clipped to ``cluster_bonus_clip``.

        Abstains (returns 0.0) when:
        - ``baseline`` is None (cluster had insufficient sample),
        - ``cluster_size < self.min_cluster_size`` (floor),
        - hunk has 0 ``function_definition`` nodes (shape undefined).

        Ramp: ``min(clip, max(0.0, |tail_z| - 1.0))`` — two-sided,
        linearly rises from 0 at 1σ to clip at (1 + clip)σ.
        """
        if baseline is None or cluster_size < self.min_cluster_size:
            return 0.0
        # Try the language inferred from the corpus parser; fall back to the
        # other grammar so the primitive works on mixed-language hunks.
        hunk_avg = _score_hunk_avg(hunk_content)
        if hunk_avg is None:
            return 0.0
        tail_z = (hunk_avg - baseline.mean) / max(baseline.std, 1e-6)
        return min(self.cluster_bonus_clip, max(0.0, abs(tail_z) - 1.0))


def _score_hunk_avg(hunk_content: str) -> float | None:
    """Try python then typescript grammar; return per-function avg or None."""
    for language in ("python", "typescript"):
        avg = _file_avg_guards(hunk_content, language)  # type: ignore[arg-type]
        if avg is not None:
            return avg
    return None


__all__ = ["FallThroughGuards"]

File: crates/argot-core/src/scoring/shape_primitives/fall_through_guards.rs
//! Fall-through-guard count — port of
//! `engine/argot/scoring/scorers/fall_through_guards.py`.
//!
//! Per function (`function_definition` / `function_declaration`), count
//! `if_statement` nodes whose `start_byte` is strictly less than the earliest
//! `return_statement`'s `start_byte` (0 if the function has no return).
//! Per-file mean of per-function counts; two-sided tail-z ramp.

use crate::scoring::adapters::Language;
use crate::scoring::shape_primitive::{Baseline, ShapePrimitive};
use crate::scoring::shape_primitives::{parse, population_mean_std};
use std::path::PathBuf;
use tree_sitter::Node;

const PY_FUNC: &str = "function_definition";
const TS_FUNC: &str = "function_declaration";
// Go: closest general case is the top-level `function_declaration` (methods
// aren't covered — this primitive is default-off in production).
const GO_FUNC: &str = "function_declaration";
const IF: &str = "if_statement";
const RETURN: &str = "return_statement";

const MIN_VALID_FILES: usize = 3;

/// Count guard `if`s before the first `return` in `func`'s subtree.
fn guards_before_return(func: Node) -> usize {
    let mut first_return: Option<usize> = None;
    let mut if_bytes: Vec<usize> = Vec::new();
    let mut stack = vec![func];
    while let Some(node) = stack.pop() {
        let kind = node.kind();
        if kind == RETURN {
            let b = node.start_byte();
            first_return = Some(match first_return {
                Some(fb) => fb.min(b),
                None => b,
            });
        } else if kind == IF {
            if_bytes.push(node.start_byte());
        }
        let n = node.child_count();
        for i in (0..n).rev() {
            if let Some(c) = node.child(i) {
                stack.push(c);
            }
        }
    }
    match first_return {
        None => 0,
        Some(fb) => if_bytes.iter().filter(|&&b| b < fb).count(),
    }
}

/// Mean guard count per function for `source`, or `None` if no functions.
fn file_avg_guards(source: &str, language: Language) -> Option<f64> {
    let tree = parse(source, language)?;
    let func_type = match language {
        Language::Python => PY_FUNC,
        Language::Typescript => TS_FUNC,
        Language::Go => GO_FUNC,
    };
    let mut counts: Vec<f64> = Vec::new();
    let mut stack = vec![tree.root_node()];
    while let Some(node) = stack.pop() {
        if node.kind() == func_type {
            counts.push(guards_before_return(node) as f64);
        }
        let n = node.child_count();
        for i in (0..n).rev() {
            if let Some(c) = node.child(i) {
                stack.push(c);
            }
        }
    }
    if counts.is_empty() {
        return None;
    }
    Some(counts.iter().sum::<f64>() / counts.len() as f64)
}

/// Try Python grammar then TypeScript; first defined average wins.
fn score_hunk_avg(hunk: &str) -> Option<f64> {
    file_avg_guards(hunk, Language::Python).or_else(|| file_avg_guards(hunk, Language::Typescript))
}

/// Fall-through-guard count primitive.
#[derive(Default)]
pub struct FallThroughGuards;

impl ShapePrimitive for FallThroughGuards {
    fn name(&self) -> &str {
        "fall_through_guards"
    }
    fn min_cluster_size(&self) -> usize {
        10
    }
    fn cluster_bonus_clip(&self) -> f64 {
        5.0
    }

    fn fit_cluster_baseline(
        &self,
        cluster_files: &[(PathBuf, String)],
        language: Language,
    ) -> Option<Baseline> {
        let mut avgs: Vec<f64> = Vec::new();
        for (_path, source) in cluster_files {
            if let Some(avg) = file_avg_guards(source, language) {
                avgs.push(avg);
            }
        }
        if avgs.len() < MIN_VALID_FILES {
            return None;
        }
        let (mean, std) = population_mean_std(&avgs);
        Some(Baseline::MeanStd { mean, std })
    }

    fn score(&self, hunk: &str, baseline: Option<&Baseline>, cluster_size: usize) -> f64 {
        let Some(Baseline::MeanStd { mean, std }) = baseline else {
            return 0.0;
        };
        if cluster_size < self.min_cluster_size() {
            return 0.0;
        }
        let Some(hunk_avg) = score_hunk_avg(hunk) else {
            return 0.0;
        };
        let tail_z = (hunk_avg - *mean) / std.max(1e-6);
        (tail_z.abs() - 1.0).max(0.0).min(self.cluster_bonus_clip())
    }
}
